### PR TITLE
ft: add metrics consumer

### PR DESCRIPTION
**Pull request by philipyoo**
This PR is focused on consuming the metrics from Kafka and storing to Redis.

These Redis metric keys will be exposed by the Backbeat Server/API (#102) through to-be-built routes.

### CODE REVIEW

**Comment by philipyoo** [comment]
@ironman-machine
SCALITY_INTEGRATION_BRANCH=S3C-1000-add-default-value-in-build
try

**Comment by philipyoo** [comment]
@ironman-machine
SCALITY_INTEGRATION_BRANCH=S3C-1000-add-default-value-in-build
try

**Comment by philipyoo** [comment]
@ironman-machine clean try

**Comment by philipyoo** [comment]
@ironman-machine try

**Comment by philipyoo** [comment]
@ironman-machine reset

**Comment by philipyoo** [comment]
@ironman-machine
SCALITY_INTEGRATION_BRANCH=rel/7.4
try

File: lib/MetricsConsumer.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        */
+        if (data.type === 'processed') {
+            this._sendRequest(redisKeys.opsDone, data.ops);
+            this._sendRequest(redisKeys.bytesDone, data.bytes);
+        } else if (data.type === 'queued') {
+            this._sendRequest(redisKeys.ops, data.ops);
+            this._sendRequest(redisKeys.bytes, data.bytes);
+        } else {
+            // unknown type
+            this.logger.error('unknown type field encountered in metrics '
```
Comment (rahulreddy) [inline lib/MetricsConsumer.js]: I have a suggestion for the logger. I recommend using a new request logger for every entry. The problem is that module logger lives and keeps accumulating logs forever and dumps the entire logs once a logger.error is called. Having request logger and ending it after the request is processed ensures that logger object doesn't grow forever, it also helps identifying a specific entry on why it failed with the reqUid.

File: lib/MetricsConsumer.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    start() {
+        const consumer = new BackbeatConsumer({
+            zookeeper: { connectionString: this.zkConfig.connectionString },
+            topic: this.mConfig.topic,
+            groupId: 'backbeat-metrics-group',
+            concurrency: CONCURRENCY,
+            queueProcessor: this.processKafkaEntry.bind(this),
+            fetchMaxBytes: CONSUMER_FETCH_MAX_BYTES,
+        });
+        consumer.on('error', () => {});
```
Comment (rahulreddy) [inline lib/MetricsConsumer.js]: No need to log the errors here?

Comment (philipyoo) [reply lib/MetricsConsumer.js]: Jonathan previously mentioned errors are logged in the `BackbeatConsumer` class so no need to log here unless expanding on the cb

File: lib/MetricsConsumer.js [hunk 3]
```diff
@@ -53,11 +53,12 @@ class MetricsConsumer {
     }
 
     processKafkaEntry(kafkaEntry, done) {
+        const log = this.logger.newRequestLogger();
         let data;
         try {
             data = JSON.parse(kafkaEntry.value);
         } catch (err) {
-            this.logger.error('error processing metrics entry', {
+            log.end().error('error processing metrics entry', {
```
Comment (rahulreddy) [inline lib/MetricsConsumer.js]: `log.end()` should be called always for all requests. You can use it this way
```js
try {
    data = JSON.parse(kafkaEntry.value);
} catch (err) {
   log.error('error processing metrics entry', {
       method: 'MetricsConsumer.processKafkaEntry',
                 error: err,
    });
    log.end();
    return done();
}


```

**Comment by rahulreddy** [comment]
@ironman-machine clean try

**Pull request merged**